lib/active_remote/attributes.rb

```diff
@@ -29,9 +72,163 @@ def write_attribute(name, value)
-        raise ActiveAttr::UnknownAttributeError, "unknown attribute: #{name}"
+        raise ::ActiveRemote::UnknownAttributeError, "unknown attribute: #{name}"
       end
     end
     alias_method :[]=, :write_attribute
+
+    # Read an attribute from the attributes hash
+    #
+    # @since 0.2.1
+    def attribute(name)
+      @attributes ||= {}
+      @attributes[name]
+    end
+
+    # Write an attribute to the attributes hash
+    #
+    # @since 0.2.1
+    def attribute=(name, value)
+      @attributes ||= {}
+      @attributes[name] = value
+    end
+
+    module ClassMethods
+      # Defines an attribute
+      #
+      # For each attribute that is defined, a getter and setter will be
+      # added as an instance method to the model. An
+      # {AttributeDefinition} instance will be added to result of the
+      # attributes class method.
+      #
+      # @example Define an attribute.
+      #   attribute :name
+      #
+      # @param (see AttributeDefinition#initialize)
+      #
+      # @raise [DangerousAttributeError] if the attribute name conflicts with
+      #   existing methods
+      #
+      # @return [AttributeDefinition] Attribute's definition
+      #
+      # @since 0.2.0
+      def attribute(name, options={})
+        if dangerous_attribute_method_name = dangerous_attribute?(name)
+          raise ::ActiveRemote::DangerousAttributeError, %{an attribute method named "#{dangerous_attribute_method_name}" would conflict with an existing method}
+        else
+          attribute! name, options
+        end
+      end
+
+      # Defines an attribute without checking for conflicts
+      #
+      # Allows you to define an attribute whose methods will conflict
+      # with an existing method. For example, Ruby's Timeout library
+      # adds a timeout method to Object. Attempting to define a timeout
+      # attribute using .attribute will raise a
+      # {DangerousAttributeError}, but .attribute! will not.
+      #
+      # @example Define a dangerous attribute.
+      #   attribute! :timeout
+      #
+      # @param (see AttributeDefinition#initialize)
+      #
+      # @return [AttributeDefinition] Attribute's definition
+      #
+      # @since 0.6.0
+      def attribute!(name, options={})
+        ::ActiveRemote::AttributeDefinition.new(name, options).tap do |attribute_definition|
+          attribute_name = attribute_definition.name.to_s
+          # Force active model to generate attribute methods
+          remove_instance_variable("@attribute_methods_generated") if instance_variable_defined?("@attribute_methods_generated")
+          define_attribute_methods([attribute_definition.name]) unless attribute_names.include? attribute_name
+          attributes[attribute_name] = attribute_definition
+        end
+      end
+
+      # Returns an Array of attribute names as Strings
+      #
+      # @example Get attribute names
+      #   Person.attribute_names
+      #
+      def attribute_names
+        attributes.keys
```

Comment:
I updated the 3.0 roadmap to make sure we addres any performance issues with the `write_attribute` method.